app/controllers/application_controller.rb

```diff
@@ -264,7 +264,12 @@ def mfa_policy
   end
 
   def aal3_policy
-    @aal3 ||= AAL3Policy.new(session: session, user: current_user)
+    @aal3 ||= AAL3Policy.new(
+      user: current_user,
+      service_provider: sp_from_sp_session,
+      auth_method: user_session[:auth_method],
```

Comment:
should we add some sort of fallback here for the old persisted params? So this doesn't mess up during a deploy? Or is the result a small enough difference that it would probably be ok? (old code key was `mfa_device_remembered`)

Something like..


```ruby
auth_method = user_session[:auth_method]
# fallback for old session data keys
if !auth_method && user_session[:mfa_device_remembered]
  auth_method = 'remember_device'
end
```

Reply:
Let me take another look and think through how this will work. Previously this would have gotten `nil` if the user had used remember device so I think it should behave appropriately. There may be other places that act wacky though.

Reply:
I went through and sorted this out. It will take several releases to do this cleanly. I am gonna open a few draft PRs for those changes and I added comments here to highlight them.

Reply:
See #3980 and #3981

Reply:
sweet thanks for planning those out